<a href="https://colab.research.google.com/github/LegendSeyi/Thai_Road_Accident_Data_Analysis/blob/main/Thai_Road_Accident_data_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# THAI ROAD ACCIDENT ANALYSIS
#### _By Adeyemi Oluwaseyi Emmanuel_

[Click here to check my Linkedin profile](https://www.linkedin.com/in/oluwaseyi-adeyemi-33b1ab197/)

[Twitter link](https://twitter.com/AmLegendseyi)

## Introduction

This dataset provides comprehensive statistics on recorded road accidents in Thailand, spanning from approximately 2019 to 2022. The data was sourced from raw information provided by the Office of the Permanent Secretary, Ministry of Transport, which is also utilized in this public dashboard for easier access and visualization. The dataset encompasses various aspects of road accidents and aims to shed light on the trends and patterns within this critical area of concern, analysis of this data could be crucial in guiding road safety policies and measures👍.

We import our required libraries

In [128]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [129]:
## creating a dictionary in python and converting it to a dataframe
tb = {'Column':['acc_code',
    'incident_datetime',
    'incident_datetime',
    'report_datetime',
    'province_th',
    'province_en',
    'agency',
    'route',
    'vehicle_type',
    'presumed_cause',
    'accident_type',
    'number_of_vehicles_involved',
    'number_of_fatalities',
    'number_of_injuries',
    'weather_condition',
    'latitude',
    'longitude',
    'road_description',
    'slope_description'],
    'Description':['The accident code or identifier',
    'The date and time of the accident occurrence',
    'The date and time when the accident was reported',
    'The name of the province in Thailand, written in Thai',
    'The name of the province in Thailand, written in English',
    'The government agency responsible for the road and traffic management',
    'The route or road segment where the accident occurred',
    'The type of vehicle involved in the accident',
    'The presumed cause or reason for the accident',
    'The type or nature of the accident',
    'The number of vehicles involved in the accident',
    'The number of fatalities resulting from the accident',
    'The number of injuries resulting from the accident',
    'The weather condition at the time of the accident',
    'The weather condition at the time of the accident',
    'The latitude coordinate of the accident location',
    'The longitude coordinate of the accident location',
    'The description of the road type or configuration where the accident occurred',
    'The description of the slope condition at the accident location']
     }

In [130]:
## Converting the dictionary to a data frame
dico = pd.DataFrame(tb)

In [131]:
dico

Column  \
0                      acc_code   
1             incident_datetime   
2             incident_datetime   
3               report_datetime   
4                   province_th   
5                   province_en   
6                        agency   
7                         route   
8                  vehicle_type   
9                presumed_cause   
10                accident_type   
11  number_of_vehicles_involved   
12         number_of_fatalities   
13           number_of_injuries   
14            weather_condition   
15                     latitude   
16                    longitude   
17             road_description   
18            slope_description   

                                          Description  
0                     The accident code or identifier  
1        The date and time of the accident occurrence  
2    The date and time when the accident was reported  
3   The name of the province in Thailand, written ...  
4   The name of the province in Thailand, written ...  
5   The government agency responsible for the road...  
6   The route or road segment where the accident o...  
7        The type of vehicle involved in the accident  
8       The presumed cause or reason for the accident  
9                  The type or nature of the accident  
10    The number of vehicles involved in the accident  
11  The number of fatalities resulting from the ac...  
12  The number of injuries resulting from the acci...  
13  The weather condition at the time of the accident  
14  The weather condition at the time of the accident  
15   The latitude coordinate of the accident location  
16  The longitude coordinate of the accident location  
17  The description of the road type or configurat...  
18  The description of the slope condition at the ...

The above table is the describtion of each columns in the dataset....

With this information, we can now import our daraset for the analysis

## IMPORTING THE DATA SET

In [132]:
# We will use the pandas library to import our dataset

In [133]:
# We can import our dataset from github


url = 'https://raw.githubusercontent.com/LegendSeyi/Dataset/main/thai_road_accident_2019_2022.csv'

pd.set_option('display.max_columns', None)
data = pd.read_csv(url, encoding= 'unicode_escape')


In [134]:
data.head()

acc_code    incident_datetime      report_datetime  \
0       571905  2019-01-01 00:00:00  2019-01-02 06:11:00   
1      3790870  2019-01-01 00:03:00  2020-02-20 13:48:00   
2       599075  2019-01-01 00:05:00  2019-01-01 10:35:00   
3       571924  2019-01-01 00:20:00  2019-01-02 05:12:00   
4       599523  2019-01-01 00:25:00  2019-01-04 09:42:00   

                                     province_th          province_en  \
0                             ลพบุรี               Loburi   
1              อุบลราชธานี     Ubon Ratchathani   
2  ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                    เชียงใหม่           Chiang Mai   
4                    นครสวรรค์         Nakhon Sawan   

                      agency  \
0  department of rural roads   
1     department of highways   
2     department of highways   
3  department of rural roads   
4     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเà...             motorcycle   
1  เดชอุดม - อุบลราชà...  private/passenger car   
2  ปราณบุรี - ปากน้ำà...             motorcycle   
3  เชื่อมทางหลวงหมà...             motorcycle   
4         เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

We have imported our data, so we can check the size of the dataset

In [135]:
data.shape

(81735, 18)

We have 81735 row and 18 columns in the dataset

In [136]:
data.columns

Index(['acc_code', 'incident_datetime', 'report_datetime', 'province_th',
       'province_en', 'agency', 'route', 'vehicle_type', 'presumed_cause',
       'accident_type', 'number_of_vehicles_involved', 'number_of_fatalities',
       'number_of_injuries', 'weather_condition', 'latitude', 'longitude',
       'road_description', 'slope_description'],
      dtype='object')

we generated the list of columns available, but out first column is not properly named..

In [137]:
# we rename the first coloumn from ''acc_code' to 'acc_code'
data.columns = [col.replace("", "") for col in data.columns]

In [138]:
data.head()

acc_code    incident_datetime      report_datetime  \
0    571905  2019-01-01 00:00:00  2019-01-02 06:11:00   
1   3790870  2019-01-01 00:03:00  2020-02-20 13:48:00   
2    599075  2019-01-01 00:05:00  2019-01-01 10:35:00   
3    571924  2019-01-01 00:20:00  2019-01-02 05:12:00   
4    599523  2019-01-01 00:25:00  2019-01-04 09:42:00   

                                     province_th          province_en  \
0                             ลพบุรี               Loburi   
1              อุบลราชธานี     Ubon Ratchathani   
2  ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                    เชียงใหม่           Chiang Mai   
4                    นครสวรรค์         Nakhon Sawan   

                      agency  \
0  department of rural roads   
1     department of highways   
2     department of highways   
3  department of rural roads   
4     department of highways   

                                               route           vehicle_type  \
0  แยกทางหลวงหมายเà...             motorcycle   
1  เดชอุดม - อุบลราชà...  private/passenger car   
2  ปราณบุรี - ปากน้ำà...             motorcycle   
3  เชื่อมทางหลวงหมà...             motorcycle   
4         เกยไชย - ศรีมงคล  private/passenger car   

                                  presumed_cause  \
0         driving under the influence of alcohol   
1                                       speeding   
2                                       speeding   
3         driving under the influence of alcohol   
4  cutting in closely by people/vehicles/animals   

                        accident_type  number_of_vehicles_involved  \
0                               other                            1   
1    rollover/fallen on straight road                            1   
2  head-on collision (not overtaking)                            2   
3                               other                            1   
4    rollover/fallen on straight road                            1   

   number_of_fatalities  number_of_injuries weather_condition   latitude  \
0                     0                   2             clear  14.959105   
1                     0                   2             clear  15.210738   
2                     1                   0             clear  12.374259   
3                     0                   1             clear  18.601721   
4                     0                   0             clear  15.866389   

    longitude road_description slope_description  
0  100.873463    straight road          no slope  
1  104.862689    straight road          no slope  
2   99.907949       wide curve        slope area  
3   98.804204    straight road          no slope  
4  100.590010    straight road          no slope

So we have successfully renamed the column

SO we want to generate the info of the dataset

In [139]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 81735 entries, 0 to 81734
Data columns (total 18 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   acc_code                     81735 non-null  int64  
 1   incident_datetime            81735 non-null  object 
 2   report_datetime              81735 non-null  object 
 3   province_th                  81735 non-null  object 
 4   province_en                  81735 non-null  object 
 5   agency                       81735 non-null  object 
 6   route                        81735 non-null  object 
 7   vehicle_type                 81735 non-null  object 
 8   presumed_cause               81735 non-null  object 
 9   accident_type                81735 non-null  object 
 10  number_of_vehicles_involved  81735 non-null  int64  
 11  number_of_fatalities         81735 non-null  int64  
 12  number_of_injuries           81735 non-null  int64  
 13  weather_conditio

We can observe that we have different data types such ad float, int, object

In [140]:
# code to generate the number of null values present in the columns
data.isnull().sum()

acc_code                         0
incident_datetime                0
report_datetime                  0
province_th                      0
province_en                      0
agency                           0
route                            0
vehicle_type                     0
presumed_cause                   0
accident_type                    0
number_of_vehicles_involved      0
number_of_fatalities             0
number_of_injuries               0
weather_condition                0
latitude                       359
longitude                      359
road_description                 0
slope_description                0
dtype: int64

we can observe that column latitude and longitude contain the same number of missing values

## ANALYSING EACH COLUMN

In [141]:
# We start by analysing the 'incident datetime' column , we convert the type of column into a datetime
data['incident_datetime'] = pd.to_datetime(data['incident_datetime'])

In [142]:
data['incident_datetime'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 81735 entries, 0 to 81734
Series name: incident_datetime
Non-Null Count  Dtype         
--------------  -----         
81735 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 638.7 KB


The column has been converted to a datetime column. so lets check if there are no null values in the dataset

In [143]:
data['incident_datetime'].isnull().sum()

0

We can confirm that there are no null values.

We can sort the data set with the incident datetime column

In [144]:
data.sort_values(by='incident_datetime', inplace=True)

In [145]:
data

acc_code   incident_datetime      report_datetime  \
0        571905 2019-01-01 00:00:00  2019-01-02 06:11:00   
1       3790870 2019-01-01 00:03:00  2020-02-20 13:48:00   
2        599075 2019-01-01 00:05:00  2019-01-01 10:35:00   
3        571924 2019-01-01 00:20:00  2019-01-02 05:12:00   
4        599523 2019-01-01 00:25:00  2019-01-04 09:42:00   
...         ...                 ...                  ...   
81730   7534454 2022-12-31 23:11:00  2023-01-01 21:49:00   
81731   7534701 2022-12-31 23:30:00  2023-01-01 22:55:00   
81732   7567022 2022-12-31 23:30:00  2023-01-23 09:17:00   
81733   7531271 2022-12-31 23:45:00  2023-01-01 05:36:00   
81734   7556715 2022-12-31 23:55:00  2023-01-11 14:50:00   

                                         province_th          province_en  \
0                                 ลพบุรี               Loburi   
1                  อุบลราชธานี     Ubon Ratchathani   
2      ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                        เชียงใหม่           Chiang Mai   
4                        นครสวรรค์         Nakhon Sawan   
...                                              ...                  ...   
81730                          นนทบุรี           Nonthaburi   
81731                                   ตราด                 Trat   
81732                    มหาสารคาม        Maha Sarakham   
81733                    กำแพงเพชร       Kamphaeng Phet   
81734                       เชียงราย           Chiang Rai   

                          agency  \
0      department of rural roads   
1         department of highways   
2         department of highways   
3      department of rural roads   
4         department of highways   
...                          ...   
81730     department of highways   
81731     department of highways   
81732     department of highways   
81733     department of highways   
81734     department of highways   

                                                   route  \
0      แยกทางหลวงหมายเà...   
1      เดชอุดม - อุบลราชà...   
2      ปราณบุรี - ปากน้ำà...   
3      เชื่อมทางหลวงหมà...   
4             เกยไชย - ศรีมงคล   
...                                                  ...   
81730  แคราย - คลองบ้านใà...   
81731               ตราด - แหลมงอบ   
81732  วาปีปทุม - พยัคฆภà...   
81733         นครชุม - มอเจริญ   
81734  แยกแม่กรณ์ - ห้วยà...   

                vehicle_type                                 presumed_cause  \
0                 motorcycle         driving under the influence of alcohol   
1      private/passenger car                                       speeding   
2                 motorcycle                                       speeding   
3                 motorcycle         driving under the influence of alcohol   
4      private/passenger car  cutting in closely by people/vehicles/animals   
...                      ...                                            ...   
81730             motorcycle                                       speeding   
81731             motorcycle                                       speeding   
81732                    van                                       speeding   
81733             motorcycle         driving under the influence of alcohol   
81734  private/passenger car                                       speeding   

                            accident_type  number_of_vehicles_involved  \
0                                   other                            1   
1        rollover/fallen on straight road                            1   
2      head-on collision (not overta

Now we can be sure that the dataset has been sorted

In [146]:
data.iloc[:,1]

0       2019-01-01 00:00:00
1       2019-01-01 00:03:00
2       2019-01-01 00:05:00
3       2019-01-01 00:20:00
4       2019-01-01 00:25:00
                ...        
81730   2022-12-31 23:11:00
81731   2022-12-31 23:30:00
81732   2022-12-31 23:30:00
81733   2022-12-31 23:45:00
81734   2022-12-31 23:55:00
Name: incident_datetime, Length: 81735, dtype: datetime64[ns]

In [147]:
data['Incident_year'] = data['incident_datetime'].astype(str).str[:4]

In [148]:
data['Incident_year'] = data['Incident_year'].astype(int)

In [149]:
data['Incident_year'].info()

<class 'pandas.core.series.Series'>
Int64Index: 81735 entries, 0 to 81734
Series name: Incident_year
Non-Null Count  Dtype
--------------  -----
81735 non-null  int64
dtypes: int64(1)
memory usage: 1.2 MB


Now We want to shift the incident_year column from the last to the second column

In [150]:
# we use the pop function
first_col = data.pop('Incident_year')

In [151]:
# then we insert it the column number position, the name of the columns, and the coloumn we popped
data.insert(1, 'Incident_year', first_col)

In [152]:
data.columns

Index(['acc_code', 'Incident_year', 'incident_datetime', 'report_datetime',
       'province_th', 'province_en', 'agency', 'route', 'vehicle_type',
       'presumed_cause', 'accident_type', 'number_of_vehicles_involved',
       'number_of_fatalities', 'number_of_injuries', 'weather_condition',
       'latitude', 'longitude', 'road_description', 'slope_description'],
      dtype='object')

In [153]:
data

acc_code  Incident_year   incident_datetime      report_datetime  \
0        571905           2019 2019-01-01 00:00:00  2019-01-02 06:11:00   
1       3790870           2019 2019-01-01 00:03:00  2020-02-20 13:48:00   
2        599075           2019 2019-01-01 00:05:00  2019-01-01 10:35:00   
3        571924           2019 2019-01-01 00:20:00  2019-01-02 05:12:00   
4        599523           2019 2019-01-01 00:25:00  2019-01-04 09:42:00   
...         ...            ...                 ...                  ...   
81730   7534454           2022 2022-12-31 23:11:00  2023-01-01 21:49:00   
81731   7534701           2022 2022-12-31 23:30:00  2023-01-01 22:55:00   
81732   7567022           2022 2022-12-31 23:30:00  2023-01-23 09:17:00   
81733   7531271           2022 2022-12-31 23:45:00  2023-01-01 05:36:00   
81734   7556715           2022 2022-12-31 23:55:00  2023-01-11 14:50:00   

                                         province_th          province_en  \
0                                 ลพบุรี               Loburi   
1                  อุบลราชธานี     Ubon Ratchathani   
2      ประจวบคีรีขันธ์  Prachuap Khiri Khan   
3                        เชียงใหม่           Chiang Mai   
4                        นครสวรรค์         Nakhon Sawan   
...                                              ...                  ...   
81730                          นนทบุรี           Nonthaburi   
81731                                   ตราด                 Trat   
81732                    มหาสารคาม        Maha Sarakham   
81733                    กำแพงเพชร       Kamphaeng Phet   
81734                       เชียงราย           Chiang Rai   

                          agency  \
0      department of rural roads   
1         department of highways   
2         department of highways   
3      department of rural roads   
4         department of highways   
...                          ...   
81730     department of highways   
81731     department of highways   
81732     department of highways   
81733     department of highways   
81734     department of highways   

                                                   route  \
0      แยกทางหลวงหมายเà...   
1      เดชอุดม - อุบลราชà...   
2      ปราณบุรี - ปากน้ำà...   
3      เชื่อมทางหลวงหมà...   
4             เกยไชย - ศรีมงคล   
...                                                  ...   
81730  แคราย - คลองบ้านใà...   
81731               ตราด - แหลมงอบ   
81732  วาปีปทุม - พยัคฆภà...   
81733         นครชุม - มอเจริญ   
81734  แยกแม่กรณ์ - ห้วยà...   

                vehicle_type                                 presumed_cause  \
0                 motorcycle         driving under the influence of alcohol   
1      private/passenger car                                       speeding   
2                 motorcycle                                       speeding   
3                 motorcycle         driving under the influence of alcohol   
4      private/passenger car  cutting in closely by people/vehicles/animals   
...                      ...                                            ...   
81730             motorcycle                                       speeding   
81731             motorcycle                                       speeding   
81732                    van                                       speeding   
81733             motorcycle         driving under the influence of alcohol   
81734  private/passenger car                                       speeding   

                            accident_type  number_of_vehicles_involved  \
0   